### PR TITLE
Prepare partial properties feature for merge

### PR DESCRIPTION
**Pull request by RikkiGibson**
#73090 
Introduces several links to tracking issue #73772
VS test insertion: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/554460
Baseline insertion for comparison: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/554589

Related spec change: dotnet/csharplang#8169

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:1111]: @cston This comment and the below assert are likely relevant for the field keyword feature.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:180]: @cston I was anticipating that perhaps the synthesized BackingField symbol will be shared between partial property parts, or otherwise the two field symbols will reference each other similarly as the associated partial property symbols. But, this work won't happen until field keyword feature is implemented, and how specifically it will be done is TBD.

File: src/Compilers/Test/Core/Assert/AssertEx.cs
Comment (RikkiGibson) [inline src/Compilers/Test/Core/Assert/AssertEx.cs:659]: Revered to main version

**Comment by RikkiGibson** [comment]
IDE layer changes are comment-only @Cosifne. These comments are also enumerated in #73772.

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:3735]: nit: indentation

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4606]: Unless there is some particular time pressure, I would favor spending a couple of days closing the remaining compiler questions/issues over punting them to a follow-up issue. They seem small and it will bring peace of mind (hopefully we don't have to revisit) and it leaves the code with fewer comments for follow-up work.

Remaining suggested tests from this morning:
- which signature do we emit in case of parameter name differences, xml docs referencing parameter names 
- SkipLocalsInit
- `new` modifier?
- split declaration (one has get, other has set) with one implementation

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4606]: If you feel like more time is needed to reach the right level of quality, that's OK. However, it is a feature I am very eager to get into our partners hands, so that they can adopt it and ship their updated generators with a decent amount of time for .NET 9. Also, I am going to be out next week, so won't be able to assist with merging the feature or getting it merge-ready during that time.

re: coverage.
- which parameter names we use in emit is tested in `IndexerParameterNameDifference`.
- xml docs parameter name difference is tested in `DocumentationCommentCompilerTests.PartialIndexer_Paramref_03` and `PartialIndexer_Paramref_04`.
- SkipLocalsInit test was added in `AttributeTests_WellKnownAttributes.SkipLocalsInit_PartialPropertyAccessor_ContainsLocalFunction`.
- `Semantics_ExtendedModifier` and `ExtendedDifference_01` tests `new`.
- Adding another split declaration test `DuplicateDeclaration_08_Definition`.
- Adding the modopt test `CustomModifierCopyTests.TestPartialPropertyOverrideCombinations`

**Comment by RikkiGibson** [comment]
I'll have to investigate the following test failure in the UsedAssemblies leg in more depth.

```log
[xUnit.net 00:00:16.14]     Microsoft.CodeAnalysis.CSharp.UnitTests.CustomModifierCopyTests.TestPartialPropertyOverrideCombinations [FAIL]
  Failed Microsoft.CodeAnalysis.CSharp.UnitTests.CustomModifierCopyTests.TestPartialPropertyOverrideCombinations [52 ms]
  Error Message:
   System.InvalidOperationException : Assertion failed
  Stack Trace:
     at Microsoft.CodeAnalysis.ThrowingTraceListener.Fail(String message, String detailMessage) in /_/src/Compilers/Test/Core/ThrowingTraceListener.cs:line 26
   at System.Diagnostics.TraceInternal.Fail(String message, String detailMessage)
   at System.Diagnostics.Debug.Fail(String message, String detailMessage)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionImpl(BoundExpression node) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:line 280
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:line 214
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitReturnStatement(BoundReturnStatement node) in /_/src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs:line 11274
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitReturnStatement(BoundReturnStatement node) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ReturnStatement.cs:line 15
   at Microsoft.CodeAnalysis.CSharp.BoundReturnStatement.Accept(BoundTreeVisitor visitor) in /_/src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs:line 3551
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitStatement(BoundStatement node) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:line 241
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitPossibleUsingDeclaration(BoundStatement node, ImmutableArray`1 statements, Int32 statementIndex, Boolean& replacedLocalDeclarations) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Block.cs:line 123
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitStatementSubList(ArrayBuilder`1 builder, ImmutableArray`1 statements, Int32 startIndex) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Block.cs:line 81
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitBlock(BoundBlock node) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Block.cs:line 33
   at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in /_/src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs:line 3293
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitStatement(BoundStatement node) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:line 241
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Rewrite(CSharpCompilation compilation, MethodSymbol method, Int32 methodOrdinal, NamedTypeSymbol containingType, BoundStatement statement, TypeCompilationState compilationState, SynthesizedSubmissionFields previousSubmissionFields, Boolean allowOmissionOfConditionalCalls, MethodInstrumentation instrumentation, DebugDocumentProvider debugDocumentProvider, BindingDiagnosticBag diagnostics, ImmutableArray`1& codeCoverageSpans, Boolean& sawLambdas, Boolean& sawLocalFunctions, Boolean& sawAwaitInExceptionHandler) in /_/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:line 146
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, MethodInstrumentation instrumentation, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& codeCoverageSpans, BindingDiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 lambdaRuntimeRudeEditsBuilder, ArrayBuilder`1 closureDebugInfoBuilder, ArrayBuilder`1 stateMachineStateDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt) in /_/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:line 1373
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) in /_/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:line 1159
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) in /_/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:line 530
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass25_0.<CompileNamedTypeAsync>b__0() in /_/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:line 432
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() in /_/src/Compilers/Core/Portable/InternalUtilities/UICultureUtilities.cs:line 139
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers() in /_/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:line 331
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean hasDeclarationErrors, Boolean emitMethodBodies, BindingDiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:line 160
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnosticsForAllMethodBodies(BindingDiagnosticBag diagnostics, Boolean doLowering, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:line 3030
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetCompleteSetOfUsedAssemblies(CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Compilation_UsedAssemblies.cs:line 77
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetUsedAssemblyReferences(CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Compilation_UsedAssemblies.cs:line 25
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.VerifyUsedAssemblyReferences(Func`1 createCompilationLambda) in /_/src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:line 1312
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.ValidateCompilation(Func`1 createCompilationLambda) in /_/src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:line 1300
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CreateCompilationCore(CSharpTestSource source, IEnumerable`1 references, CSharpCompilationOptions options, CSharpParseOptions parseOptions, String assemblyName, String sourceFileName, Boolean skipUsesIsNullable, Nullable`1 experimentalFeature, Boolean skipExtraValidation) in /_/src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:line 1275
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CreateEmptyCompilation(CSharpTestSource source, IEnumerable`1 references, CSharpCompilationOptions options, CSharpParseOptions parseOptions, String assemblyName, String sourceFileName, Boolean skipUsesIsNullable, Boolean skipExtraValidation) in /_/src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:line 1239
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CreateCompilation(CSharpTestSource source, IEnumerable`1 references, CSharpCompilationOptions options, CSharpParseOptions parseOptions, TargetFramework targetFramework, String assemblyName, String sourceFileName, Boolean skipUsesIsNullable) in /_/src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:line 1228
   at Microsoft.CodeAnalysis.CSharp.UnitTests.CustomModifierCopyTests.TestPartialPropertyOverrideCombinations() in /_/src/Compilers/CSharp/Test/Symbol/Symbols/Source/CustomModifierCopyTests.cs:line 344
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4606]: Thanks for the coverage update
I think it's better to wait until week of June 10th to button up and merge. But I'm okay if Jared is okay with merging with those known follow-ups. Then I would recommend we do the follow-up soon while things are "loaded" in our minds and preferably before 17.11 ships.

**Review by jcouv — APPROVED** [review]
Compiler changes LGTM Thanks (iteration 9)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/CustomModifierCopyTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/CustomModifierCopyTests.cs:276]: I didn't see any similar test to this already so figured we might as well have it.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/CustomModifierCopyTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/CustomModifierCopyTests.cs:394]: Had to change the test away from using `[]` due to #73804.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 10)

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4606]: Discussed offline, our plan is to merge to main after this PR is merged and to address the remaining items in #73772 in 17.11

File: src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs
Comment (Cosifne) [inline src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs:93]: 🙂Then we would need a new 'AbstractPartialPropertiesProvider'

File: src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs
Comment (Cosifne) [inline src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs:829]: This means if the renaming symbol  is a parameter of a partial method, make sure rename engine also checks the method for conflict.
I don't think this would apply to partial property. 
The only similar case would be if the property is  a indexer, but in that case the propertie would be `this[int bar]`. So still don't need to check if `this` would case conflict

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/OrdinaryMethodReferenceFinder.cs
Comment (Cosifne) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/OrdinaryMethodReferenceFinder.cs:37]: Yeah this is MethodReferenceFinder, do not need to check this.
Maybe should check this in PropertyReferenceFinder

File: src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs
Comment (RikkiGibson) [reply src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs:93]:
```suggestion
        // https://github.com/dotnet/roslyn/issues/73772: should we also add an `AbstractPartialPropertiesProvider`?
```

File: src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs
Comment (RikkiGibson) [reply src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs:829]:
```suggestion
```

**Review by Cosifne — APPROVED** [review]

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:1691]: >*definition* part parameters are used when analyzing implementations. [](http://example.com/codeflow?start=45&length=69)

Is this statement correct? If we used the definition part parameters for analyzing the implementation, why are reporting warnings `// 3` and `// 4`?

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:3784]: >} [](http://example.com/codeflow?start=8&length=1)

Are we binding the initializers in these error cases? Consider adding a test such as:
```csharp
partial string P4 { get; set; } = nameof(Unknown);
partial string P4 { get => ""; set { } }
```

**Comment by cston** [comment]
>                     public partial int this[int p2] { get => p2; set { } }

Consider using `p2` in the `set` accessor as well: `set { _ = p2; }`.

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4589 in 63e1536. [](commit_id = 63e1536217cd8f7f80a46b0d7f9e1a78cc48d9c0, deletion_comment = False)

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4610]: >} [](http://example.com/codeflow?start=8&length=1)

Would it make sense to test which parameter names are in scope? For instance:
```csharp
class A : System.Attribute
{
    public A(object x, object y) { }
}
partial class C
{
    public partial int this[int p1] { [param: A(nameof(p1), nameof(p2))] set; }
    public partial int this[int p2] { [param: A(nameof(p1), nameof(p2))] set { } }
}
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:1691]: You're right, I'll adjust the comment and the spec PR. Don't know how I got mixed up on this.

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:1693]:
```suggestion
            // The implementation part signature is used to analyze the implementation part bodies.
            // The definition part signature is used to analyze use sites.
```

**Review by cston — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs:4713]: I know why this is happening. The implementation part owns binding attributes on parameters.

https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/extended-nameof-scope.md doesn't appear to spec what should happen here.

I am hesitant to even spec the behavior here (which is already shipped for methods).

**Pull request merged**
